src/high-level/matrix.lisp

```diff
@@ -0,0 +1,557 @@
+;;;; matrix.lisp
+;;;;
+;;;; Author: Cole Scott
+;;;;         Robert Smith
+
+(in-package #:magicl)
+
+(deftype matrix-storage (&optional type)
+  `(simple-array ,type (*)))
+
+(defstruct (matrix (:include abstract-tensor)
+                   (:constructor nil)
+                   (:copier nil))
+  (nrows 0 :type alexandria:non-negative-fixnum)
+  (ncols 0 :type alexandria:non-negative-fixnum)
+  (size 0 :type alexandria:non-negative-fixnum :read-only t)
+  (layout :column-major :type (member :row-major :column-major)))
+
+(defmethod nrows ((m matrix))
+  (matrix-nrows m))
+
+(defmethod ncols ((m matrix))
+  (matrix-ncols m))
+
+(defmethod size ((m matrix))
+  (matrix-size m))
+
+(defmethod layout ((m matrix))
+  (matrix-layout m))
+
+;;; Specfic matrix classes
+(defmacro defmatrix (name type tensor-class)
+  "Define a new matrix subclass with the specified NAME and element TYPE,
+compatible with TENSOR-CLASS, as well as the abstract-tensor methods
+required not specified by the generic MATRIX class (MAKE-TENSOR,
+ELEMENT-TYPE, CAST, COPY-TENSOR, DEEP-COPY-TENSOR, TREF, SETF TREF)"
+  (let ((constructor-sym (intern (format nil "MAKE-~a" name)))
+        (copy-sym (intern (format nil "COPY-~a" name)))
+        (storage-sym (intern (format nil "~a-STORAGE" name))))
```

Comment:
Replace `name` by `(string-upcase (symbol-name name))` to avoid problems when `*print-case*` is set to `:downcase`.